### PR TITLE
Add 0.4.21 CHANGELOG

### PR DESCRIPTION
**Pull request by Kubuxu**
WIP. Please either add or mention in comments what should be featured.

### CODE REVIEW

**Comment by Stebalien** [comment]
> Should we clean up dependency changelog?

Probably. Really, I wouldn't print out the version changes, that doesn't say much. Instead, I'd print the actual change log (commit messages) from each dep in one of our orgs (multiformats, ipfs, ipld, libp2p, etc.).

Going through that also helps flesh out the features/changes.

**Comment by Kubuxu** [comment]
I can write that, but this will be even more noise.
We for sure don't want to include git changelogs for the whole libp2p ecosystem in our changelog.

**Comment by djdv** [comment]
minor change (may or may not be worth mentioning) in: https://github.com/ipfs/go-ipfs/pull/6277
>We're now base64 encoding values returned by dht get when encoding them into JSON

Previously these where byte slices converted to strings which could produce invalid values/output.

**Comment by Stebalien** [comment]
>  We for sure don't want to include git changelogs for the whole libp2p ecosystem in our changelog.

It's actually quite useful as we benefit from all of those changes. We don't _have_ to include them but I wouldn't include the version changes regardless.

Take a look at the changelogs from the last few releases.

> minor change (may or may not be worth mentioning) in: #6277

We should definitely mention that as a breaking change.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,278 @@
...
+Go-ipfs will accept TLS 1.3 handshakes but will still primarily use secio.
+If you want to help testing new handshake, you can enable [experiment](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#tls-13-as-default-handshake-protocol).
+
+### New: Build system `GOCC` variable
+
+Build system now uses `GOCC` environment variable allowing for use of multiple
+go versions during builds.
+
+
+(please point out what should be featured, doesn't have to be go-ipfs).
```
Comment (momack2) [inline CHANGELOG.md]: Should we pull the list of contributors like Alan did in the most recent js-ipfs release? https://github.com/ipfs/js-ipfs/releases/tag/v0.36.0

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,278 @@
...
+#### Fixed: `ipfs add` with multiple files
+
+In version 0.4.20 a bug was introduced preventing `ipfs add fileA fileB` from working.
+This was fixed in ([ipfs/go-ipfs#6255](https://github.com/ipfs/go-ipfs/pull/6255)).
+
+### New: TLS 1.3 handshake
+
+Go-libp2p has gained ability to use TLS 1.3 for connection encryption and authentication.
+Go-ipfs will accept TLS 1.3 handshakes but will still primarily use secio.
+If you want to help testing new handshake, you can enable [experiment](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#tls-13-as-default-handshake-protocol).
```
Comment (momack2) [inline CHANGELOG.md]:
```suggestion
If you want to help testing the new handshake, you can enable [this experiment](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#tls-13-as-default-handshake-protocol).
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,278 @@
...
+  - add an experiment to prefer TLS 1.3 over secio ([ipfs/go-ipfs#6229](https://github.com/ipfs/go-ipfs/pull/6229))
+  - build: fix macos build with fuse ([ipfs/go-ipfs#6235](https://github.com/ipfs/go-ipfs/pull/6235))
+  - dep: update libp2p, datastore, and the dht ([ipfs/go-ipfs#6240](https://github.com/ipfs/go-ipfs/pull/6240))
+  - Support --human flag in cmd/object-stat ([ipfs/go-ipfs#6241](https://github.com/ipfs/go-ipfs/pull/6241))
+  - constructor: cleanup some things ([ipfs/go-ipfs#6246](https://github.com/ipfs/go-ipfs/pull/6246))
+  - core/corehttp/gateway_handler: pass a request ctx instead of the node ([ipfs/go-ipfs#6244](https://github.com/ipfs/go-ipfs/pull/6244))
+  - reprovider: Use goprocess ([ipfs/go-ipfs#6248](https://github.com/ipfs/go-ipfs/pull/6248))
+  -  add: Fix adding multiple files ([ipfs/go-ipfs#6255](https://github.com/ipfs/go-ipfs/pull/6255))
+  - dep: update libp2p, dht, peerstore, swarm, secio ([ipfs/go-ipfs#6257](https://github.com/ipfs/go-ipfs/pull/6257))
+  - Fix AutoNAT service for private network ([ipfs/go-ipfs#6251](https://github.com/ipfs/go-ipfs/pull/6251))
```
Comment (momack2) [inline CHANGELOG.md]: is this notable?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,278 @@
...
+  - dep: update go-ipfs-config ([ipfs/go-ipfs#6285](https://github.com/ipfs/go-ipfs/pull/6285))
+  - Test and fix GC/pin bug ([ipfs/go-ipfs#6288](https://github.com/ipfs/go-ipfs/pull/6288))
+  - core: fix autonat construction ([ipfs/go-ipfs#6289](https://github.com/ipfs/go-ipfs/pull/6289))
+  - commands(pin update): return resolved CIDs instead of paths ([ipfs/go-ipfs#6275](https://github.com/ipfs/go-ipfs/pull/6275))
+  - Go test fix ([ipfs/go-ipfs#6293](https://github.com/ipfs/go-ipfs/pull/6293))
+  - License update to dual MIT and Apache 2 ([ipfs/go-ipfs#6301](https://github.com/ipfs/go-ipfs/pull/6301))
+  - docs: fix Routing section ([ipfs/go-ipfs#6309](https://github.com/ipfs/go-ipfs/pull/6309))
+  - fix(Dockerfile): Allow ipfs mount in Docker container ([ipfs/go-ipfs#5560](https://github.com/ipfs/go-ipfs/pull/5560))
+  - test(coreapi): use a thread-safe datastore everywhere ([ipfs/go-ipfs#6222](https://github.com/ipfs/go-ipfs/pull/6222))
+  - ping: fix a bunch of issues ([ipfs/go-ipfs#6312](https://github.com/ipfs/go-ipfs/pull/6312))
```
Comment (momack2) [inline CHANGELOG.md]: any of these worth calling out?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,278 @@
...
+  - fail start when loading a plugin fails ([ipfs/go-ipfs#6339](https://github.com/ipfs/go-ipfs/pull/6339))
+  - fix formatting in object get ([ipfs/go-ipfs#6340](https://github.com/ipfs/go-ipfs/pull/6340))
+  - docs+mk: update guidance for unsupported platforms ([ipfs/go-ipfs#6338](https://github.com/ipfs/go-ipfs/pull/6338))
+  - Add golangci-lint ([ipfs/go-ipfs#6321](https://github.com/ipfs/go-ipfs/pull/6321))
+  - readme: remove mention of DCO ([ipfs/go-ipfs#6344](https://github.com/ipfs/go-ipfs/pull/6344))
+  - Add generation of dependency changes to mkreleaselog ([ipfs/go-ipfs#6348](https://github.com/ipfs/go-ipfs/pull/6348))
+  - doc: remove out of date documentation ([ipfs/go-ipfs#6345](https://github.com/ipfs/go-ipfs/pull/6345))
+  - core: Use correct default for connmgr lowWater ([ipfs/go-ipfs#6352](https://github.com/ipfs/go-ipfs/pull/6352))
+  - fix vendoring dependencies when building the source tarball ([ipfs/go-ipfs#6349](https://github.com/ipfs/go-ipfs/pull/6349))
+  - feat: improve errors when a path fails to parse ([ipfs/go-ipfs#6346](https://github.com/ipfs/go-ipfs/pull/6346))
```
Comment (momack2) [inline CHANGELOG.md]: yay for better errors?

File: CHANGELOG.md (hunk 3 above)
Comment (Stebalien) [reply CHANGELOG.md]: It's unlikely to affect many users.

File: CHANGELOG.md (hunk 4 above)
Comment (Stebalien) [reply CHANGELOG.md]: Done.

File: CHANGELOG.md (hunk 5 above)
Comment (Stebalien) [reply CHANGELOG.md]: Done.

**Comment by Stebalien** [comment]
I've filled out the release notes and added a recursive changelog.

**Comment by Stebalien** [comment]
cc @ipfs/wg-go-core please review.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,412 @@
...
+   completely emptied.
+2. It was loading every provider record into memory whenever we updated the set
+   of providers.
+
+Combined, these two issues were trashing the provider record cache, forcing the
+DHT to repeatedly load and discard provider records.
+
+#### Improved Connection Management
+
+Go-ipfs uses a service called the "connection manager" to rank connections them
```
Comment (djdv) [inline CHANGELOG.md]: >to rank connections them by "usefulness" and close the least-useful ones when go-ipfs collects too many.

"them" needs to be removed
>to rank connections by "usefulness" and close the least-useful ones when go-ipfs collects too many.

**Comment by eingenito** [comment]
@stebalien Should we mention the intention to default to dhtclient routing in a future release?

**Comment by Stebalien** [comment]
> @Stebalien Should we mention the intention to default to dhtclient routing in a future release?

Thanks for the reminder. Yes!

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,412 @@
...
+
+#### [BREAKING] Base32 Encoded v1 CIDs By Default
+
+Both js-ipfs and go-ipfs now encode CIDv1 CIDs using base32 by default, instead
+of base58. Unfortunately, base58 is case-sensitive and doesn't play well with
+browsers (see [#4143](https://github.com/ipfs/go-ipfs/issues/4143).
+
+#### Human Readable Numbers
+
+The `ipfs bitswap stat` and and `ipfs object stat` commands now support a
```
Comment (michaelavila) [inline CHANGELOG.md]:
```suggestion
The `ipfs bitswap stat` and `ipfs object stat` commands now support a
```

**Comment by Stebalien** [comment]
> @Stebalien Should we mention the intention to default to dhtclient routing in a future release?

Ugh. It looks like configs specify `"dht"` by default, not `""`. We'd have to reinterpret that to change the default for existing nodes.

We can still make it the default for future nodes but that doesn't really require an announcement this release.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,412 @@
...
+`--humanize` flag that formats numbers with human-readable units (GiB, MiB,
+etc.).
+
+#### Improved Errors
+
+This release improves to types of errors:
```
Comment (michaelavila) [inline CHANGELOG.md]:
```suggestion
This release improves two types of errors:
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,412 @@
...
+
+```bash
+> make build GOFLAGS=-tags=openssl
+```
+
+### CoreAPI
+
+The CoreAPI refactor is still underway and we've made significant progress
+towards a usable ipfs-as-a-library constructor. Specifically, we've integrated
+integrating the [fx](https://go.uber.org/fx) dependency injection system and are
```
Comment (michaelavila) [inline CHANGELOG.md]:
```suggestion
the [fx](https://go.uber.org/fx) dependency injection system and are
```

File: CHANGELOG.md (hunk 8 above)
Comment (Stebalien) [reply CHANGELOG.md]: :man_facepalming:

**Comment by Stebalien** [comment]
We can clean these up post-release if necessary. Let's get this out the door.

**Pull request merged**